64_FFT/main.cpp

```diff
@@ -0,0 +1,326 @@
+			return false;
+
+		// this time we load a shader directly from a file
+		smart_refctd_ptr<IGPUShader> shader;
+		{
+			IAssetLoader::SAssetLoadParams lp = {};
+			lp.logger = m_logger.get();
+			lp.workingDirectory = ""; // virtual root
+			auto assetBundle = m_assetMgr->getAsset("app_resources/shader.comp.hlsl", lp);
+			const auto assets = assetBundle.getContents();
+			if (assets.empty())
+				return logFail("Could not load shader!");
+
+			// lets go straight from ICPUSpecializedShader to IGPUSpecializedShader
+			auto source = IAsset::castDown<ICPUShader>(assets[0]);
+			// The down-cast should not fail!
+			assert(source);
+
+			// this time we skip the use of the asset converter since the ICPUShader->IGPUShader path is quick and simple
+			shader = m_device->createShader(source.get());
+			if (!shader)
+				return logFail("Creation of a GPU Shader to from CPU Shader source failed!");
+		}
+
+		// The StreamingTransientDataBuffers are actually composed on top of another useful utility called `CAsyncSingleBufferSubAllocator`
+		// The difference is that the streaming ones are made on top of ranges of `IGPUBuffer`s backed by mappable memory, whereas the
+		// `CAsyncSingleBufferSubAllocator` just allows you suballocate subranges of any `IGPUBuffer` range with deferred/latched frees.
+		constexpr uint32_t DownstreamBufferSize = sizeof(output_t) << 23;
+		constexpr uint32_t UpstreamBufferSize = sizeof(input_t) << 23;
+
+		m_utils = make_smart_refctd_ptr<IUtilities>(smart_refctd_ptr(m_device), smart_refctd_ptr(m_logger), DownstreamBufferSize, UpstreamBufferSize);
+		if (!m_utils)
+			return logFail("Failed to create Utilities!");
+		m_upStreamingBuffer = m_utils->getDefaultUpStreamingBuffer();
+		m_downStreamingBuffer = m_utils->getDefaultDownStreamingBuffer();
+		m_upStreamingBufferAddress = m_upStreamingBuffer->getBuffer()->getDeviceAddress();
+		m_downStreamingBufferAddress = m_downStreamingBuffer->getBuffer()->getDeviceAddress();
+
+		// People love Reflection but I prefer Shader Sources instead!
+		const nbl::asset::SPushConstantRange pcRange = { .stageFlags = IShader::ESS_COMPUTE,.offset = 0,.size = sizeof(PushConstantData) };
+
+		// This time we'll have no Descriptor Sets or Layouts because our workload has a widely varying size
+		// and using traditional SSBO bindings would force us to update the Descriptor Set every frame.
+		// I even started writing this sample with the use of Dynamic SSBOs, however the length of the buffer range is not dynamic
+		// only the offset. This means that we'd have to write the "worst case" length into the descriptor set binding.
+		// Then this has a knock-on effect that we couldn't allocate closer to the end of the streaming buffer than the "worst case" size.
+		{
+			auto layout = m_device->createPipelineLayout({ &pcRange,1 });
+			IGPUComputePipeline::SCreationParams params = {};
+			params.layout = layout.get();
+			params.shader.shader = shader.get();
```

Comment:
super important nitpick.

If you expect the subgroups to "properly" fill out a workgroup and not get packed haphazardly, you need to set `computeFullSubgroups` to true.

You might also benefit (perf wise) from setting the subgroup size to the max subgroup size possible.

Reply:
TODO, full subgroups are super important!

Reply:
THIS IS NEEEDED!